### PR TITLE
Organize functions into functions module

### PR DESCRIPTION
**Pull request by Y-Nak**
Ref: #6680 
Rearrange lints in `functions`.

changelog: none

### CODE REVIEW

File: clippy_lints/src/functions/mod.rs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    /// }
+    /// ```
+    pub TOO_MANY_LINES,
+    pedantic,
+    "functions with too many lines"
+}
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for public functions that dereference raw pointer
+    /// arguments but are not marked unsafe.
```
Comment (giraffate) [inline clippy_lints/src/functions/mod.rs:59]: nits
```suggestion
    /// arguments but are not marked `unsafe`.
```

File: clippy_lints/src/functions/mod.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    /// ```
+    pub MUST_USE_CANDIDATE,
+    pedantic,
+    "function or method that could take a `#[must_use]` attribute"
+}
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for public functions that return a `Result`
+    /// with an `Err` type of `()`. It suggests using a custom type that
+    /// implements [`std::error::Error`].
```
Comment (giraffate) [inline clippy_lints/src/functions/mod.rs:168]: nits, it seems to have no link, so it would be better to remove `[]`
```suggestion
    /// implements `std::error::Error`.
```

File: clippy_lints/src/functions/result_unit_err.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        let err_ty = substs.type_at(1);
+        if err_ty.is_unit();
+        then {
+            span_lint_and_help(
+                cx,
+                RESULT_UNIT_ERR,
+                fn_header_span,
+                "this returns a `Result<_, ()>",
+                None,
+                "use a custom Error type instead",
```
Comment (giraffate) [inline clippy_lints/src/functions/result_unit_err.rs:62]: nits, it might be better to fix this in another PR because the stderr file needs to be fixed.
```suggestion
                "this returns a `Result<_, ()>`",
                None,
                "use a custom `Error` type instead",
```

**Review by giraffate — COMMENTED** [review]
Overall seems good, thanks! I made some comments.

File: clippy_lints/src/functions/result_unit_err.rs (hunk 3 above)
Comment (Y-Nak) [reply clippy_lints/src/functions/result_unit_err.rs:62]: Ok, I'll address this in another PR.

**Comment by Y-Nak** [comment]
Changes:
Tweak documents in `functions`.

**Review by giraffate — APPROVED** [review]

**Comment by giraffate** [comment]
@bors r+

It looks good, thanks!
